internal/reports/report_v2.go

```diff
@@ -0,0 +1,319 @@
+		statementStats = append(statementStats, StatementStat{Statement: s, Schema: x.Schema, Data: x.Data, Skip: x.Skip, Error: x.Error})
+	}
+	return statementStats
+}
+
+func fetchNameChanges(conv *internal.Conv) (nameChanges []NameChange) {
+	for tableId, spTable := range conv.SpSchema {
+		srcTable := conv.SrcSchema[tableId]
+		if srcTable.Name != spTable.Name {
+			nameChanges = append(nameChanges, NameChange{NameChangeType: "TableName", SourceTable: srcTable.Name, OldName: srcTable.Name, NewName: spTable.Name})
+		}
+		for colId, spCol := range spTable.ColDefs {
+			srcCol, ok := srcTable.ColDefs[colId]
+			if !ok {
+				continue
+			}
+			if srcCol.Name != spCol.Name {
+				nameChanges = append(nameChanges, NameChange{NameChangeType: "ColumnName", SourceTable: srcTable.Name, OldName: srcCol.Name, NewName: spCol.Name})
+			}
+		}
+		for _, spFk := range conv.SpSchema[tableId].ForeignKeys {
+			srcFk, err := internal.GetSrcFkFromId(conv.SrcSchema[tableId].ForeignKeys, spFk.Id)
+			if err != nil {
+				continue
+			}
+			if srcFk.Name != spFk.Name {
+				nameChanges = append(nameChanges, NameChange{NameChangeType: "ForeignKey", SourceTable: srcTable.Name, OldName: srcFk.Name, NewName: spFk.Name})
+			}
+		}
+		for _, spIdx := range conv.SpSchema[tableId].Indexes {
+			srcIdx, err := internal.GetSrcIndexFromId(conv.SrcSchema[tableId].Indexes, spIdx.Id)
+			if err != nil {
+				continue
+			}
+			if srcIdx.Name != spIdx.Name {
+				nameChanges = append(nameChanges, NameChange{NameChangeType: "Index", SourceTable: srcTable.Name, OldName: srcIdx.Name, NewName: spIdx.Name})
+			}
+		}
+	}
+	return nameChanges
+}
+
+func fetchTableReports(inputTableReports []tableReport, conv *internal.Conv) (tableReports []TableReport) {
+	for _, t := range inputTableReports {
+		//1. src and Sp Table Names
+		tableReport := TableReport{SrcTableName: conv.SrcSchema[t.SrcTable].Name}
+		tableReport.SpTableName = conv.SrcSchema[t.SrcTable].Name
+
+		//2. Schema Report
+		migrationType := *conv.Audit.MigrationType
+		if migrationType != migration.MigrationData_DATA_ONLY {
+			tableReport.SchemaReport = getSchemaReport(t.Cols, t.Warnings, t.SyntheticPKey != "")
+		}
+		//3. Data Report
+		schemaOnly := conv.SchemaMode()
+		if !schemaOnly {
+			tableReport.DataReport = getDataReport(t.rows, t.badRows, conv.Audit.DryRun)
+		}
+		//4. Warnings
+		for _, x := range t.Body {
+			var warnings = Warnings{WarningType: x.Heading}
+			for _, l := range x.Lines {
+				warnings.WarningList = append(warnings.WarningList, l)
+			}
+			tableReport.Warnings = append(tableReport.Warnings, warnings)
+		}
+		tableReports = append(tableReports, tableReport)
+	}
+	return tableReports
+}
+
+func getSchemaReport(cols, warnings int64, missingPKey bool) (schemaReport SchemaReport) {
+	schemaReport.TotalColumns = cols
+	schemaReport.Warnings = warnings
+	schemaReport.PkMissing = missingPKey
+	switch {
+	case cols == 0:
+		schemaReport.Rating = "NONE"
+	case warnings == 0:
```

Comment:
Are we getting away with considering missing PK that we do for text reports while deciding on rating of schema migration ? And will it not lead to inconsistency in the report

Reply:
Earlier we used to append a static text to the rating if the PK was missing. In the structured report, this is captured in a flag `missingPKey`.
In the text report, `missingPKey` is used to generate the static add-on text ("with PK missing").

Reply:
Other than the add-on text, rating also differs in the report text file hence the discrepancy issue.

Reply:
Thanks, I had duplicated the scoring logic for both schema and data in two places. `RateSchema` method contains the actual logic for scoring (which accounts for `missingPKey`). I have replaced this switch-case with that method call for both schema and data now.